File: 19.state/state.java
// Estado
interface State {
    void insertMoney();
    void dispenseProduct();
}

// Estado concreto: Lista
class ReadyState implements State {
    private VendingMachine machine;

    public ReadyState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertMoney() {
        System.out.println("Dinero insertado. Selecciona un producto.");
        machine.setState(machine.getWaitingState());
    }

    @Override
    public void dispenseProduct() {
        System.out.println("Inserta dinero primero.");
    }
}

// Estado concreto: Esperando dinero
class WaitingState implements State {
    private VendingMachine machine;

    public WaitingState(VendingMachine machine) {
        this.machine = machine;
    }

    @Override
    public void insertMoney() {
        System.out.println("Dinero ya insertado. Selecciona un producto.");
    }

    @Override
    public void dispenseProduct() {
        System.out.println("Producto dispensado. Gracias por tu compra.");
        machine.setState(machine.getReadyState());
    }
}

// Contexto
class VendingMachine {
    private State readyState;
    private State waitingState;
    private State currentState;

    public VendingMachine() {
        readyState = new ReadyState(this);
        waitingState = new WaitingState(this);
        currentState = readyState;
    }

    public void setState(State state) {
        currentState = state;
    }

    public State getReadyState() {
        return readyState;
    }

    public State getWaitingState() {
        return waitingState;
    }

    public void insertMoney() {
        currentState.insertMoney();
    }

    public void dispenseProduct() {
        currentState.dispenseProduct();
    }
}

// Uso
public class state {
    public static void main(String[] args) {
        VendingMachine machine = new VendingMachine();

        machine.insertMoney();
        machine.dispenseProduct();
        machine.dispenseProduct(); // Comportamiento según el estado actual
    }
}

File: 19.state/state.ts
// Estado
interface State {
    insertMoney(): void;
    dispenseProduct(): void;
}

// Estado concreto: Lista
class ReadyState implements State {
    constructor(private machine: VendingMachine) {}

    public insertMoney(): void {
        console.log("Dinero insertado. Selecciona un producto.");
        this.machine.setState(this.machine.getWaitingState());
    }

    public dispenseProduct(): void {
        console.log("Inserta dinero primero.");
    }
}

// Estado concreto: Esperando dinero
class WaitingState implements State {
    constructor(private machine: VendingMachine) {}

    public insertMoney(): void {
        console.log("Dinero ya insertado. Selecciona un producto.");
    }

    public dispenseProduct(): void {
        console.log("Producto dispensado. Gracias por tu compra.");
        this.machine.setState(this.machine.getReadyState());
    }
}

// Contexto
class VendingMachine {
    private readyState: State;
    private waitingState: State;
    private currentState: State;

    constructor() {
        this.readyState = new ReadyState(this);
        this.waitingState = new WaitingState(this);
        this.currentState = this.readyState;
    }

    public setState(state: State): void {
        this.currentState = state;
    }

    public getReadyState(): State {
        return this.readyState;
    }

    public getWaitingState(): State {
        return this.waitingState;
    }

    public insertMoney(): void {
        this.currentState.insertMoney();
    }

    public dispenseProduct(): void {
        this.currentState.dispenseProduct();
    }
}

// Uso
const machine = new VendingMachine();

machine.insertMoney();
machine.dispenseProduct();
machine.dispenseProduct(); // Comportamiento según el estado actual